include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp

```diff
@@ -0,0 +1,376 @@
+
+/*!\file
+ * \brief Provides seqan3::detail::two_dimensional_matrix.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include <seqan3/alignment/matrix/detail/matrix_coordinate.hpp>
+#include <seqan3/alignment/matrix/detail/two_dimensional_matrix_iterator_base.hpp>
+#include <seqan3/core/type_traits/deferred_crtp_base.hpp>
+#include <seqan3/core/type_traits/template_inspection.hpp>
+#include <seqan3/std/algorithm>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+//!\brief Strong type for setting the column dimension of a matrix.
+//!\ingroup alignment_matrix
+struct number_cols : strong_type<size_t, number_cols>
+{
+    //!\brief Import the base constructors.
+    using strong_type<size_t, number_cols>::strong_type;
+};
+
+//!\brief Strong type for setting the row dimension of a matrix.
+//!\ingroup alignment_matrix
+struct number_rows : strong_type<size_t, number_rows>
+{
+    //!\brief Import the base constructors.
+    using strong_type<size_t, number_rows>::strong_type;
+};
+
+/*!\brief A two dimensional matrix used inside of alignment algorithms.
+ * \ingroup alignment_matrix
+ * \implements std::ranges::RandomAccessRange
+ *
+ * \tparam value_t The value type to store.
+ * \tparam allocator_t The allocator type used to allocate the storage; defaults to std::allocator.
+ * \tparam order The matrix-major-order; defaults to seqan3::detail::matrix_major_order::row.
+ *
+ * \details
+ *
+ * This two dimensional matrix type is a base data structure for several alignment matrices. It can be customised by
+ * providing another allocator type. Depending on the given `order` the element access on the underlying buffer
+ * follows a row-major-order or a column-major-order.
+ */
+template <typename value_t,
+          typename allocator_t = std::allocator<value_t>,
+          matrix_major_order order = matrix_major_order::row>
+class two_dimensional_matrix
+{
+private:
+
+    /*!\name Auxiliary types
+     * \{
+     */
+    using storage_type = std::vector<value_t, allocator_t>; //!< The storage type.
+    //!\}
+
+    /*!\brief A two-dimensional matrix iterator.
+     * \tparam matrix_t The underlying seqan3::detail::two_dimensional_matrix.
+     *
+     * \details
+     *
+     * Offers a two-dimensional iterator interface over the seqan3::detail::two_dimensional_matrix, which stores the
+     * data in a flattened one-dimensional vector.
+     */
+    template <typename matrix_t>
+    class iterator_type :
+        public two_dimensional_matrix_iterator_base<iterator_type<matrix_t>, order>
+    {
+    private:
+        //!\brief The base class type.
+        using base_t = two_dimensional_matrix_iterator_base<iterator_type<matrix_t>, order>;
+
+        //!\brief Befriend the base crtp class.
+        template <typename derived_t, matrix_major_order other_order>
+            requires is_type_specialisation_of_v<derived_t, iterator_type> && other_order == order
```

Comment:
I'm not sure, but I think requires in friend declaration is not allowed.

Reply:
Mhmm, why should it not. It constraints the template arguments so a friend declaration shouldn't make an exception

Reply:
I'm not completely sure, but I guess that there is no wording in the standard for this and practically is a forward declaration. And also haven't seen any rules for forward declarations.

Reply:
Same with me. I couldn't find any documentation against it. So I assume as a template expression I can apply a requires clause.